File: rust/src/ipc/contracts.rs
// #exonware/xwsystem/rust/src/ipc/contracts.rs
//! Company: eXonware.com
//! Author: Eng. Muhammad AlShehri
//! Email: connect@exonware.com
//! Version: 0.1.0.1
//! Generation Date: September 04, 2025
//! 
//! IPC module contracts - interfaces and enums for inter-process communication.


use std::collections::HashMap;

use crate::defs::{IPCType, MessageQueueType, MessageType, ProcessState, QueueType, SharedMemoryType};

// Aliases for backward compatibility

/// Interface for message queue operations.
pub trait IMessageQueue {
    /// Put message in queue.
    async fn put(&self, message: serde_json::Value, message_type: MessageType) -> ();

    /// Get message from queue.
    async fn get(&self, timeout: Option<f64>) -> serde_json::Value;

    /// Get message without waiting.
    async fn get_nowait(&self) -> serde_json::Value;

    /// Check if queue is empty.
    fn empty(&self) -> bool;

    /// Check if queue is full.
    fn full(&self) -> bool;

    /// Get queue size.
    fn size(&self) -> i64;

}

/// Interface for pipe operations.
pub trait IPipe {
    /// Send data through pipe.
    async fn send(&self, data: serde_json::Value) -> ();

    /// Receive data from pipe.
    async fn recv(&self, timeout: Option<f64>) -> serde_json::Value;

    /// Receive data without waiting.
    async fn recv_nowait(&self) -> serde_json::Value;

    /// Close pipe.
    fn close(&self) -> ();

    /// Check if pipe is closed.
    fn closed(&self) -> bool;

}

/// Interface for shared memory operations.
pub trait ISharedMemory {
    /// Create shared memory segment.
    fn create(&self, name: String, size: i64) -> ();

    /// Attach to shared memory segment.
    fn attach(&self, name: String) -> ();

    /// Detach from shared memory segment.
    fn detach(&self) -> ();

    /// Unlink shared memory segment.
    fn unlink(&self) -> ();

    /// Read from shared memory.
    fn read(&self, offset: i64, size: Option<i64>) -> Vec<u8>;

    /// Write to shared memory.
    fn write(&self, data: Vec<u8>, offset: i64) -> ();

}

/// Interface for process management.
pub trait IProcessManager {
    /// Create new process.
    fn create_process(&self, target: fn(), args: (), kwargs: Option<HashMap<String, serde_json::Value>>) -> Process;

    /// Start process.
    fn start_process(&self, process: Process) -> ();

    /// Stop process.
    fn stop_process(&self, process: Process, timeout: Option<f64>) -> ();

    /// Get process state.
    fn get_process_state(&self, process: Process) -> ProcessState;

    /// Check if process is alive.
    fn is_process_alive(&self, process: Process) -> bool;

}

/// Interface for process pool operations.
pub trait IProcessPool {
    /// Submit task to pool.
    fn submit(&self, func: fn()) -> serde_json::Value;

    /// Submit async task to pool.
    async fn submit_async(&self, func: fn()) -> serde_json::Value;

    /// Map function over iterable.
    fn map(&self, func: fn(), iterable: Vec<serde_json::Value>) -> Vec<serde_json::Value>;

    /// Map function over iterable asynchronously.
    async fn map_async(&self, func: fn(), iterable: Vec<serde_json::Value>) -> Vec<serde_json::Value>;

    /// Close process pool.
    fn close(&self) -> ();

    /// Join all processes.
    fn join(&self) -> ();

}

File: src/exonware/xwsystem/ipc/contracts.py
#exonware/xwsystem/src/exonware/xwsystem/ipc/contracts.py
"""
Company: eXonware.com
Author: eXonware Backend Team
Email: connect@exonware.com
Version: 0.9.0.9
Generation Date: September 04, 2025
IPC module contracts - interfaces and enums for inter-process communication.
"""

from typing import Any, Protocol, runtime_checkable

from collections.abc import AsyncGenerator, Callable
from multiprocessing import Process
# Import enums from types module
from .defs import (
    IPCType,
    MessageType,
    ProcessState,
    QueueType,
    SharedMemoryType,
    MessageQueueType
)
@runtime_checkable

class IMessageQueue(Protocol):
    """Interface for message queue operations."""

    async def put(self, message: Any, message_type: MessageType = MessageType.DATA) -> None:
        """Put message in queue."""
        ...

    async def get(self, timeout: float | None = None) -> Any:
        """Get message from queue."""
        ...

    async def get_nowait(self) -> Any:
        """Get message without waiting."""
        ...

    def empty(self) -> bool:
        """Check if queue is empty."""
        ...

    def full(self) -> bool:
        """Check if queue is full."""
        ...

    def size(self) -> int:
        """Get queue size."""
        ...
@runtime_checkable

class IPipe(Protocol):
    """Interface for pipe operations."""

    async def send(self, data: Any) -> None:
        """Send data through pipe."""
        ...

    async def recv(self, timeout: float | None = None) -> Any:
        """Receive data from pipe."""
        ...

    async def recv_nowait(self) -> Any:
        """Receive data without waiting."""
        ...

    def close(self) -> None:
        """Close pipe."""
        ...

    def closed(self) -> bool:
        """Check if pipe is closed."""
        ...
@runtime_checkable

class ISharedMemory(Protocol):
    """Interface for shared memory operations."""

    def create(self, name: str, size: int) -> None:
        """Create shared memory segment."""
        ...

    def attach(self, name: str) -> None:
        """Attach to shared memory segment."""
        ...

    def detach(self) -> None:
        """Detach from shared memory segment."""
        ...

    def unlink(self) -> None:
        """Unlink shared memory segment."""
        ...

    def read(self, offset: int = 0, size: int | None = None) -> bytes:
        """Read from shared memory."""
        ...

    def write(self, data: bytes, offset: int = 0) -> None:
        """Write to shared memory."""
        ...
@runtime_checkable

class IProcessManager(Protocol):
    """Interface for process management."""

    def create_process(self, target: Callable, args: tuple = (), kwargs: dict | None = None) -> Process:
        """Create new process."""
        ...

    def start_process(self, process: Process) -> None:
        """Start process."""
        ...

    def stop_process(self, process: Process, timeout: float | None = None) -> None:
        """Stop process."""
        ...

    def get_process_state(self, process: Process) -> ProcessState:
        """Get process state."""
        ...

    def is_process_alive(self, process: Process) -> bool:
        """Check if process is alive."""
        ...
@runtime_checkable

class IProcessPool(Protocol):
    """Interface for process pool operations."""

    def submit(self, func: Callable, *args, **kwargs) -> Any:
        """Submit task to pool."""
        ...

    async def submit_async(self, func: Callable, *args, **kwargs) -> Any:
        """Submit async task to pool."""
        ...

    def map(self, func: Callable, iterable: list[Any]) -> list[Any]:
        """Map function over iterable."""
        ...

    async def map_async(self, func: Callable, iterable: list[Any]) -> list[Any]:
        """Map function over iterable asynchronously."""
        ...

    def close(self) -> None:
        """Close process pool."""
        ...

    def join(self) -> None:
        """Join all processes."""
        ...
